src/jdk.jfr/share/classes/jdk/jfr/internal/tool/Metadata.java

```diff
@@ -91,52 +102,158 @@ int groupValue(Type t) {
+            if (acceptSingleOption(options, "--ids")) {
                 showIds = true;
             }
+            if (acceptFilterOption(options, "--events")) {
+                if (foundEventFilter) {
+                    throw new UserSyntaxException("use --events event1,event2,event3 to include multiple events");
+                }
+                foundEventFilter = true;
+                String filterStr = options.remove();
+                warnForWildcardExpansion("--events", filterStr);
+                filter = addEventFilter(filterStr, filter);
+            }
+            if (acceptFilterOption(options, "--categories")) {
+                if (foundCategoryFilter) {
+                    throw new UserSyntaxException("use --categories category1,category2 to include multiple categories");
+                }
+                foundCategoryFilter = true;
+                String filterStr = options.remove();
+                warnForWildcardExpansion("--categories", filterStr);
+                filter = addCategoryFilter(filterStr, filter);
+            }
             if (optionCount == options.size()) {
                 // No progress made
+                checkCommonError(options, "--event", "--events");
+                checkCommonError(options, "--category", "--categories");
                 throw new UserSyntaxException("unknown option " + options.peek());
             }
             optionCount = options.size();
         }
 
-        try (PrintWriter pw = new PrintWriter(System.out)) {
+        try (PrintWriter pw = new PrintWriter(System.out, false, Charset.forName("UTF-8"))) {
             PrettyWriter prettyWriter = new PrettyWriter(pw);
             prettyWriter.setShowIds(showIds);
-            try (RecordingFile rf = new RecordingFile(file)) {
-                List<Type> types = PRIVATE_ACCESS.readTypes(rf);
+            if (filter != null) {
+                filter = addCache(filter, eventType -> eventType.getId());
+            }
+
+            // determine whether reading from recording file or reading from the JDK where
+            // the jfr tool is located will be used
+            List<EventType> types = null;
+            if (file != null) {
+                try (RecordingFile rf = new RecordingFile(file)) {
+                    types = rf.readEventTypes();
+                } catch (IOException ioe) {
+                     couldNotReadError(file, ioe);
+                }
+            } else {
+                // FlightRecorder.getEventTypes returns unmodifiable list thus disallowing sorting
+                // so copy its elements to a new list and sort it
+                types = new ArrayList<>(FlightRecorder.getFlightRecorder().getEventTypes());
+            }
+            if (types != null) {
                 Collections.sort(types, new TypeComparator());
-                for (Type type : types) {
-                    prettyWriter.printType(type);
+            }
+            for (EventType type : types) {
```

Comment:
This looks like a change of behavior, the metadata command no longer prints all the types, only event types.

Reply:
`FlightRecorder.getEventTypes()` only returns event types:
```
$ jfr metadata | grep ^class
class IntFlagChanged extends jdk.jfr.Event {
class JVMInformation extends jdk.jfr.Event {
class JavaErrorThrow extends jdk.jfr.Event {
...
```
Should we use `jdk.jfr.internal.TypeLIbrary.getTypes()`? As you mentioned earlier, this would introduces the `jdk.jfr.internal` package, it returns all types:
```
$ jfr metadata dummy.jfr | grep ^class
class boolean {
class byte {
class char {
class double {
class float {
class int {
class long {
class short {
class Class {
class String {
class Thread {
class CalleeMethod {
class ClassLoader {
class BooleanFlag extends java.lang.annotation.Annotation {
class Category extends java.lang.annotation.Annotation {
class ContentType extends java.lang.annotation.Annotation {
class IntFlagChanged extends jdk.jfr.Event {
class JVMInformation extends jdk.jfr.Event {
class JavaErrorThrow extends jdk.jfr.Event {
...
```
But as a drawback, we can not reuse existing category&event filter and related methods since they are only suitable for event types.

Reply:
The comment about FlightRecorder.getEventTypes() was on the test. If possible, I think we should avoid depending on internal packages in tests, as it makes them more fragile. Product code can depend on internal packages.

Reply:
Changed.